text/0000-custom-panic-handlers.md

```diff
@@ -0,0 +1,125 @@
+The `debug` handler would be the default and have a behavior similar to the current `on_panic`, to avoid modifying the behavior of existing code.
+`basic` would print whatever is passed to `panic!()` without adding information to the message.
+`silent` would never output anything. 
+
+A handler (or a *callback*, as this was called in `rt::unwind`'s code) is registered with `thread::on_panic`. It has the following signature:
+
+``` rust
+fn(msg: &Any + Send, file: &'static str, line: u32) -> bool
+```
+
+It is almost the same as the current [`Callback`](http://doc.rust-lang.org/std/rt/unwind/type.Callback.html) signature, except that it returns a `bool` to indicate whether handler execution should go on after the execution of the callee. 
+
+Handlers that simply add a layer of logging return `true` to execute the next one in the list, and 'overriding' handlers return `false` to stop there.  
+Handlers are called in reverse order of registration (first registered, last called), to place the 'most important' ones (usually one of `thread::handlers`) at the top.  
+
+
+This new API would be used as follows:
+
+```rust
+use std::thread;
+
+fn main() {
+    thread::spawn(|| {
+        // use the default panic handler, `debug`, the one we currently have
+
+        some_optional.unwrap();
+    }).join();
+
+    thread::spawn(|| {
+        // Always use the `silent` handler
+        thread::on_panic(thread::handlers::silent);
+
+        // This will never display a message
+        
+        // Use cases:
+        // - abort thread silently on errors unrecoverable at its level,
+        //   but recoverable at the process level, and spawn a new one
+        a_third_party_function_that_could_fail();
+    }).join();
+
+    thread::spawn(|| {
+        // Always use the `basic` handler
+        thread::on_panic(thread::handlers::basic);
+
+        // This will simply display whatever is passed to `panic!` on panic,
+        // much like println!()
+        
+        // Use cases:
+        // - abort thread on invalid input with a user-friendly message
+        if !is_valid(input) {
+            panic!("Invalid input - please try again");
+        } else {
+            // ...
+        }
+        
+    }).join();
+    
+    
+}
+```
+
+Users wishing to have `debug` messages on debug builds and `basic` messages on release builds could simply do this:
+
+```rust
+thread::on_panic(
+    if cfg!(debug) {
+        thread::handlers::debug
+    } else {
+        thread::handlers::basic
+    }
+);
+```
+
+
+**The `basic` handler does *not* print a newline when `panic!` is called without arguments**, to allow silencing `panic!` in specific cases (silent recovery by spawning a new thread) without hiding other `panic` messages which may be useful for debugging.
+
+
+# Drawbacks
+
+As this doesn't break existing code, the only one I can think of is that this will require a pretty heavy refactoring of `rt::unwind`.
```

Comment:
Could you expand this drawbacks section a little more? For example the RFC as-is does not allow handlers to close over any data, requiring thread-local-storage for that sort of data. Additionally, the currently specified chaining mechanism can be seen as somewhat complicated.